sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -87,11 +87,21 @@ static RT::PiProgram createSpirvProgram(const RT::PiContext Context,
 RT::PiProgram ProgramManager::getBuiltOpenCLProgram(OSModuleHandle M,
                                                     const context &Context) {
   std::shared_ptr<context_impl> Ctx = getSyclObjImpl(Context);
-  std::map<OSModuleHandle, RT::PiProgram> &CachedPrograms =
-      Ctx->getCachedPrograms();
-  auto It = CachedPrograms.find(M);
-  if (It != CachedPrograms.end())
-    return It->second;
+
+  {
+    auto Locked = Ctx->acquireCachedPrograms();
+    auto &CachedPrograms = Locked.get();
```

Comment:
Now it's not clear what type variables have.
 LLVM coding guidelines say that auto can be used when "the type is already obvious from the context" and  "Another time when auto works well for these purposes is when the type would have been abstracted away anyways, often behind a container’s typedef such as std::vector<T>::iterator.". I suggest specifying exact type.